moveos/moveos-stdlib/rooch-framework/sources/events.move

```diff
@@ -0,0 +1,87 @@
+/// The Event module defines an `EventHandleGenerator` that is used to create
+/// `EventHandle`s with unique GUIDs. It contains a counter for the number
+/// of `EventHandle`s it generates. An `EventHandle` is used to count the number of
+/// events emitted to a handle and emit events to the event store.
+module rooch_framework::events {
+    use std::error;
+    use std::bcs;
+    use std::signer;
+    use std::vector;
+    use moveos_std::account_storage;
+    use moveos_std::storage_context::{StorageContext};
+
+    /// A resource representing the counter used to generate uniqueness under each account. There won't be destructor for
+    /// this resource to guarantee the uniqueness of the generated handle.
+    struct EventHandleGenerator has key {
+        // A monotonically increasing counter
+        counter: u64,
+        addr: address,
+    }
+
+    /// A handle for an event such that:
+    /// 1. Other modules can emit events to this handle.
+    /// 2. Storage can use this handle to prove the total number of events that happened in the past.
+    struct EventHandle<phantom T: drop + store> has store {
+        /// Total number of events emitted to this event stream.
+        counter: u64,
+        /// A globally unique ID for this event stream.
+        guid: vector<u8>,
+    }
+
+    /// The event generator resource was in an invalid state
+    const EEventGenerator: u64 = 0;
+
+    /// Publishs a new event handle generator.
+    public fun publish_generator(ctx: &mut StorageContext, account: &signer) {
+        let addr = signer::address_of(account);
+        assert!(!account_storage::global_exists<EventHandleGenerator>(ctx, addr), error::already_exists(EEventGenerator));
+        account_storage::global_move_to(
+            ctx,
+            account,
+            EventHandleGenerator{ counter: 0, addr }
+        )
+    }
```

Comment:
We do not need the `EventHandleGenerator` if we use the ObjectID as the event stream `guid`. This makes the event API simpler.

Reply:
I tried. Originally designed according to the sui event style, the contract does not contain guid at all, which is the simplest and more intuitive for developers.

However, when implementing native event, it is found that to implement an event mechanism similar to Sui, it is necessary to expand the existing Move VM virtual machine to implement a runtime event adapter. 

The current transaction is based on the native Move VM, and the transaction and event are bound, which means that to expand the Event, it needs to be connected with the transaction, and the cost of implementation is very high.

Sui achieves a similar effect based on ObjectRuntime,  Its transactions and events are generated based on ObjectRuntime.

Events are related to the following two interfaces of the Move VM core:

1. Save event in native function: 
```
context.save_event(guid, seq_num, ty, msg)
```

2. Derived events after the transaction is executed:
```
let (change_set, raw_events, mut extensions) = session.finish_with_extensions()?;
```